File: modules/communications/java/u2f-commons/src/main/java/org/crysil/communications/u2f/RegisterInternalHandler.java
package org.crysil.communications.u2f;

import org.crysil.commons.Module;
import org.crysil.communications.json.JsonUtils;
import org.crysil.communications.u2f.data.RegisterInternalRequest;
import org.crysil.communications.u2f.data.RegisterInternalResponse;
import org.crysil.logging.Logger;
import org.crysil.protocol.Response;
import org.crysil.protocol.payload.crypto.generatekey.PayloadGenerateU2FKeyResponse;
import org.crysil.protocol.payload.crypto.sign.PayloadSignResponse;

import com.fasterxml.jackson.core.util.ByteArrayBuilder;
import com.google.common.primitives.Bytes;

import static org.crysil.communications.u2f.U2FUtil.*;

public class RegisterInternalHandler implements Handler {

	private final CrySILForwarder skytrustHandler;

	public RegisterInternalHandler(CrySILForwarder skytrustHandler) {
		this.skytrustHandler = skytrustHandler;
	}

	@Override
	public String handle(String u2fRawRequest, Module actor, U2FReceiverInterface receiver) {
		RegisterInternalRequest u2fInternalRegisterRequest = JsonUtils.fromJson(u2fRawRequest,
				RegisterInternalRequest.class);
		if (u2fInternalRegisterRequest == null) {
			return null;
		}
		Logger.debug("Incoming request: " + JsonUtils.toJson(u2fInternalRegisterRequest));

		byte[] signature;
		byte[] keyEncoded;
		byte[] certEncoded;
		byte[] keyHandleBytes;
		try {
			byte[] challengeParam = u2fInternalRegisterRequest.getChallengeHash();
			byte[] appParam = u2fInternalRegisterRequest.getAppIdHash();

			Response responseGenKey = skytrustHandler.executeGenerateWrappedKey(
					u2fInternalRegisterRequest.getChallengeHash(), u2fInternalRegisterRequest.getAppIdHash(), null,
					actor, receiver);
			if (responseGenKey == null || responseGenKey.getPayload() == null) {
				Logger.error("No response for generate wrapped key");
				return null;
			}
			Logger.debug("Generated wrapped key: " + JsonUtils.toJson(responseGenKey));
			if (responseGenKey.getPayload().getType().equals("authChallengeRequest"))
				return JsonUtils.toJson(responseGenKey);

			PayloadGenerateU2FKeyResponse payloadGenKey = (PayloadGenerateU2FKeyResponse) responseGenKey.getPayload();

			javax.security.cert.X509Certificate cert = payloadGenKey.getCertificate();
			keyEncoded = cert.getPublicKey().getEncoded();
			if (keyEncoded.length > 65) {
				keyEncoded = stripMetaData(keyEncoded);
			}
			certEncoded = cert.getEncoded();
			keyHandleBytes = payloadGenKey.getEncodedRandom();

			byte[] signatureData = buildSignatureBytes(appParam, challengeParam, keyHandleBytes, keyEncoded);
			Response responseSign = skytrustHandler.executeSignatureRequest(payloadGenKey.getEncodedWrappedKey(),
					signatureData, actor, receiver);
			if (responseSign == null || responseSign.getPayload() == null) {
				Logger.error("No response for sign");
				return null;
			}
			Logger.debug("Generated signature: " + JsonUtils.toJson(responseSign));
			PayloadSignResponse payloadSign = (PayloadSignResponse) responseSign.getPayload();
			signature = payloadSign.getSignedHashes().iterator().next();
		} catch (Exception e) {
			e.printStackTrace();
			return null;
		}

		RegisterInternalResponse response;
		boolean fromU2FActor = Bytes.indexOf(signature, keyHandleBytes) > -1
				&& Bytes.indexOf(signature, certEncoded) > -1;
		if (!fromU2FActor) {
			byte[] signatureData = buildResponseBytes(keyEncoded, keyHandleBytes, certEncoded, signature);
			response = new RegisterInternalResponse(signatureData);
		} else {
			response = new RegisterInternalResponse(signature);
		}
		Logger.debug("Answering with: " + JsonUtils.toJson(response));

		return JsonUtils.toJson(response);
	}

	private static byte[] buildSignatureBytes(byte[] appParam, byte[] challengeParam, byte[] keyHandleBytes,
			byte[] keyEncoded) {
		ByteArrayBuilder builder = new ByteArrayBuilder();
		builder.write((byte) 0x00);
		builder.write(appParam);
		builder.write(challengeParam);
		builder.write(keyHandleBytes);
		builder.write(keyEncoded);
		byte[] result = builder.toByteArray();
		builder.close();
		return result;
	}

	private static byte[] buildResponseBytes(byte[] keyEncoded, byte[] keyHandleBytes, byte[] certEncoded,
			byte[] signature) {
		ByteArrayBuilder builder = new ByteArrayBuilder();
		builder.write(0x05);
		builder.write(keyEncoded);
		builder.write(keyHandleBytes.length);
		builder.write(keyHandleBytes);
		builder.write(certEncoded);
		builder.write(signature);
		byte[] result = builder.toByteArray();
		builder.close();
		return result;
	}

}

File: modules/others/windows-cp/ControlU2FKeys/RegisterInternalHandler.cs
using System;
using System.Linq;
using System.Security.Cryptography.X509Certificates;
using Newtonsoft.Json.Linq;

namespace ControlU2FKeys
{
    public class RegisterInternalHandler : IHandler
    {
        private readonly CrySilForwarder handler;

        public RegisterInternalHandler(CrySilForwarder handler)
        {
            this.handler = handler;
        }

        private byte[] StripMetaData(byte[] pubKey)
        {
            if (pubKey.Length == 65)
                return pubKey;
            if (pubKey.Length < 4)
                return pubKey;
            int headerLen = pubKey[3];
            if (pubKey.Length < headerLen + 6)
                return pubKey;
            int keyLen = pubKey[headerLen + 5];
            if (pubKey.Length < headerLen + 7 + keyLen - 1)
                return pubKey;
            byte[] key = new byte[keyLen - 1]; // strip first byte of key too
            Array.Copy(pubKey, headerLen + 7, key, 0, keyLen - 1);
            return key;
        }

        static bool BytesContains(byte[] haystack, byte[] needle)
        {
            var len = needle.Length;
            var limit = haystack.Length - len;
            for (var i = 0; i <= limit; i++)
            {
                var k = 0;
                for (; k < len; k++)
                {
                    if (needle[k] != haystack[i + k]) break;
                }
                if (k == len)
                    return true;
            }
            return false;
        }

        public string Handle(string msg)
        {
            var u2fRequest = JObject.Parse(msg);
            var challengeParam = Helpers.Base64UrlDecode(u2fRequest["challengeHash"].ToString());
            var appParam = Helpers.Base64UrlDecode(u2fRequest["appIdHash"].ToString());
            var crysilResponse = this.handler.ExecuteGenerateWrappedKey(u2fRequest["challengeHash"].ToString(), u2fRequest["appIdHash"].ToString(), null);
            var responseGenKey = JObject.Parse(crysilResponse);
            var payloadGenKey = responseGenKey["payload"];
            var certEncoded = Helpers.Base64UrlDecode(payloadGenKey["encodedX509Certificate"].ToString());
            var certificate = new X509Certificate(certEncoded);
            var keyEncoded = StripMetaData(certificate.GetPublicKey());
            var keyHandleBytes = Helpers.Base64UrlDecode(payloadGenKey["encodedRandom"].ToString());
            var signatureBytes = new byte[] { 0x00 }
                .Concat(appParam)
                .Concat(challengeParam)
                .Concat(keyHandleBytes)
                .Concat(keyEncoded)
                .ToArray();
            var responseStr = this.handler.ExecuteSignatureRequest(payloadGenKey["encodedWrappedKey"].ToString(),
                Helpers.Base64UrlEncode(signatureBytes));
            var responseSign = JObject.Parse(responseStr);
            var payloadSign = responseSign["payload"];
            var signature = Helpers.Base64UrlDecode(payloadSign["signedHashes"][0].ToString());
            var signatureData = new byte[] { 0x05 }
                .Concat(keyEncoded)
                .Concat(new byte[] { (byte)keyHandleBytes.Length })
                .Concat(keyHandleBytes)
                .Concat(certEncoded)
                .Concat(signature)
                .ToArray();
            var fromU2FActor = BytesContains(signature, keyHandleBytes) && BytesContains(signature, certEncoded);
            if (fromU2FActor)
            {
                signatureData = signature;
            }
            var response = new JObject
            {
                ["registrationData"] = Helpers.Base64UrlEncode(signatureData)
            };
            return response.ToString();
        }
    }
}